shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java

```diff
@@ -1237,25 +1238,42 @@ public boolean runDistCpWithSnapshots(String oldSnapshot, String newSnapshot, Li
-          returnCode = rDistcp.run(rParams.toArray(new String[0]));
-          if (returnCode == 0) {
-            LOG.info("Target restored to previous state.  source: {} target: {} snapshot: {}. Reattempting to copy.",
-                srcPaths, dst, oldSnapshot);
-            dst.getFileSystem(conf).deleteSnapshot(dst, oldSnapshot);
-            dst.getFileSystem(conf).createSnapshot(dst, oldSnapshot);
-            returnCode = distcp.run(params.toArray(new String[0]));
+        if (targetModified(dst, conf, oldSnapshot)) {
+          if (overwriteTarget) {
+            LOG.warn("Copy failed due to target modified. Attempting to restore back the target. source: {} target: {} "
+                    + "snapshot: {}", srcPaths, dst, oldSnapshot);
+            List<String> rParams = constructDistCpWithSnapshotParams(srcPaths, dst, ".", oldSnapshot, conf, "-rdiff");
+            DistCp rDistcp = new DistCp(conf, null);
+            returnCode = rDistcp.run(rParams.toArray(new String[0]));
             if (returnCode == 0) {
+              LOG.info("Target restored to previous state.  source: {} target: {} snapshot: {}. Reattempting to copy.",
+                      srcPaths, dst, oldSnapshot);
+              dst.getFileSystem(conf).deleteSnapshot(dst, oldSnapshot);
+              dst.getFileSystem(conf).createSnapshot(dst, oldSnapshot);
+              returnCode = distcp.run(params.toArray(new String[0]));
+              if (returnCode == 0) {
+                return true;
+              } else {
+                LOG.error("Copy failed with after target restore for source: {} to target: {} snapshot1: {} snapshot2: "
+                        + "{} params: {}. Return code: {}", srcPaths, dst, oldSnapshot, newSnapshot, params, returnCode);
+                return false;
+              }
+            }
+          } else {
+            //in case overwriteTarget is false, and we encounter an exception due to targetFs getting modified
+            // since last snapshot, then fall back to full distcp
+            LOG.warn("Copy failed due to target modified and overwrite is false. Attempting full distcp." +
+                    "Source:{}, target: {}",srcPaths, dst);
+            // Get the path relative to the initial snapshot for copy.
+            Path snapRelPath = new Path(srcPaths.get(0), HdfsConstants.DOT_SNAPSHOT_DIR + "/" + newSnapshot);
+            // Copy from the initial snapshot path.
+            if(runDistCp(Collections.singletonList(snapRelPath), dst, conf)) {
               return true;
             } else {
-              LOG.error("Copy failed with after target restore for source: {} to target: {} snapshot1: {} snapshot2: "
-                  + "{} params: {}. Return code: {}", srcPaths, dst, oldSnapshot, newSnapshot, params, returnCode);
+              LOG.error("Copy failed with full distcp for source: {} to target: {}  snapshot:"
+                      + "{} params: {}. Return code: {}", srcPaths, dst, newSnapshot, returnCode);
```

Comment:
Incorrect number of placeholders. Misses params

Reply:
Corrected